File: leetcode/typescript/000695.max-area-of-island.ts
// Time complexity: O(m * n)
// Space complexity: O(m * n)
function maxAreaOfIsland(grid: number[][]): number {
  const rows = grid.length;
  const cols = grid[0].length;
  const visited = new Set<string>();
  let maxArea = 0;

  function bfs(r: number, c: number) {
    const queue = [[r, c]];
    let area = 0;

    while (queue.length > 0) {
      const [r, c] = queue.shift()!;
      visited.add(`${r},${c}`);
      area++;

      const directions = [[-1, 0], [1, 0], [0, -1], [0, 1]];
      for (const [dr, dc] of directions) {
        const newR = r + dr;
        const newC = c + dc;

        if (
          newR < 0 || newC < 0 || newR >= rows || newC >= cols ||
          grid[newR][newC] === 0 ||
          visited.has(`${newR},${newC}`)
        ) {
          continue;
        }

        queue.push([newR, newC]);
        visited.add(`${newR},${newC}`);
      }
    }

    maxArea = Math.max(maxArea, area);
  }

  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      if (grid[r][c] !== 0 && !visited.has(`${r},${c}`)) {
        bfs(r, c);
      }
    }
  }

  return maxArea;
}

File: leetcode/python/000695.max-area-of-island.py
class Solution:
    # Time complexity: O(m * n)
    # Space complexity: O(m * n)
    def maxAreaOfIsland(self, grid: List[List[int]] -> int):
        rows, cols = len(grid), len(grid[0])
        max_area = 0
        visited = set()
        
        def dfs(r, c):
            if r < 0 or c < 0 or r >= rows or c >= cols or (r, c) in visited or grid[r][c] == 0:
                return 0

            visited.add((r, c))
            return 1 + dfs(r + 1, c) + dfs(r - 1, c) + dfs(r, c + 1) + dfs(r, c - 1)

        for r in range(rows):
            for c in range(cols):
                area = dfs(r, c)
                max_area = max(max_area, area)
                       
        return max_area
